### PR TITLE
Feature/it's classified! (Classify Actions)

### PR DESCRIPTION
**Pull request by Spedwards**
Supersedes #1033.

Created the PR mostly for tracking.

As evident by the `tower.Action` file, this will also move tower code into individual actions.

### CODE REVIEW

File: action.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    
+    getTargetByID(id) {
+        return Game.getObjectById(id) || Game.creeps[id];
+    };
+    
+    isValidAction(object) {
+        return true;
+    };
+    
+    isValidTarget(targt, object) {
```
Comment (chobobobo) [inline action.js]: typo

File: creep.Action.js [hunk 2]
```diff
@@ -1,125 +1,57 @@
...
-            var workResult = this.work(creep);
-            if( workResult != OK ) {
+        if (range <= this.targetRange) {
+            const workResult = this.work(creep);
+            if (workResult !== OK) {
                 const tryAction = creep.action;
                 const tryTarget = creep.target;
                 creep.action = null;
                 creep.target = null;
                 creep.handleError({errorCode: workResult, action: this, target: creep.target, range, creep});
```
Comment (chobobobo) [inline creep.Action.js]: creep.target will always be null, this should probably be `target: tryTarget` and `action: tryAction` right?

File: creep.Action.js [hunk 3]
```diff
@@ -1,125 +1,57 @@
...
                 return;
             }
-            range = creep.pos.getRangeTo(creep.target); // target may have changed (eg. hauler feed+move/tick)
+            range = creep.pos.getRangeTo(creep.target);
```
Comment (chobobobo) [inline creep.Action.js]: that comment explained why we are getting range twice, it avoids someone thinking they can refactor it out.

File: creep.Action.js [hunk 4]
```diff
@@ -1,125 +1,57 @@
...
-            return options || {};
-        }
+    
+    registerAction(creep, target) {
+        Population.registerAction(creep, this, target);
+        this.onAssignment(creep, target);
     };
-    // strategy accessor
-    this.selectStrategies = function() {
-        return [this.defaultStrategy];
```
Comment (chobobobo) [inline creep.Action.js]: Where have the action strategy resolution code moved to if they are removed from here?  They should be set in the constructor I believe.

File: creep.Action.js (hunk 2 above)
Comment (Spedwards) [inline creep.Action.js]: Or moved up a couple lines. Wasn't part of my changes but I'll make it one.

File: creep.Action.js (hunk 3 above)
Comment (Spedwards) [inline creep.Action.js]: Comments only got removed when I was off in another world while typing. Will try and go back and add any missing ones at the end.

File: action.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        if (target && this.isAddableTarget(object, target)) {
+            if (!object.action || object.action.name !== this.name || !object.target || object.target.id !== target.id || object.target.name !== target.name) {
+                this.registerAction(object, target);
+            }
+            return true;
+        }
+        return false;
+    };
+    
+    registerAction(object, target) {
```
Comment (karlthepagan) [inline action.js]: This creates prototype inheritance, right?

I'll be very interested in how this behaves with virals. See https://github.com/ScreepsOCS/screeps.behaviour-action-pattern/pull/413 where my request to do class like inheritance was previously rejected.

**Pull request closed without merging**